kroxylicious/src/main/java/io/kroxylicious/proxy/service/ContributionManager.java

```diff
@@ -0,0 +1,51 @@
+/*
+ * Copyright Kroxylicious Authors.
+ *
+ * Licensed under the Apache Software License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+package io.kroxylicious.proxy.service;
+
+import java.util.Map;
+import java.util.ServiceLoader;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.function.BiFunction;
+import java.util.function.Function;
+
+@SuppressWarnings({ "rawtypes", "unchecked" })
+public class ContributionManager /* <T, C extends Context, S extends Contributor<T, C>> */ {
+    public static final ContributionManager INSTANCE = new ContributionManager();
+
+    private final Map<Class, Iterable> contributors;
+    private final Function<Class, Iterable> loaderFunction;
+
+    private ContributionManager() {
+        this(ServiceLoader::load);
+    }
+
+    /* test */ ContributionManager(Function<Class, Iterable> loaderFunction) {
+        this.contributors = new ConcurrentHashMap<>();
+        this.loaderFunction = loaderFunction;
+    }
+
+    public <T, C extends Context, S extends Contributor<T, C>> ConfigurationDefinition getDefinition(Class<S> contributorClass, String typeName,
+                                                                                                     BiFunction<Class<S>, String, String> errorGenerator) {
+        return findContributor(contributorClass, typeName, (typName, contributor) -> contributor.getConfigDefinition(typName), errorGenerator);
+    }
+
+    public <T, C extends Context, S extends Contributor<T, C>> T getInstance(Class<S> contributorClass, String typeName, C constructionContext,
+                                                                             BiFunction<Class<S>, String, String> errorGenerator) {
+        return findContributor(contributorClass, typeName, (typName, contributor) -> contributor.getInstance(typeName, constructionContext), errorGenerator);
+    }
+
+    private <T, C extends Context, S extends Contributor<T, C>, X> X findContributor(Class<S> contributorClass, String typeName, BiFunction<String, S, X> extractor,
+                                                                                     BiFunction<Class<S>, String, String> errorGenerator) {
```

Comment:
Can people voice opinions on ` BiFunction<Class<S>, String, String> errorGenerator` vs `String errorMessage`. 

The `BiFunction` feels like the right API for `ContributionManager` however none of the call sites use the `class` parameter and in fact all of them know the typename anyway. The only argument I can see is for GC reasons we could switch them all from String concatenation to string format's and do the format lazily when the `errorGenerator` is invoked.

Reply:
Is it worth considering not having customizable errors? Is it not enough to have the ContributionManager throw an error with message like: `Name '${typeName}' is not contributed by any ${serviceInterface}`.

The error function makes me think "which errors will trigger this function, does it need to discriminate between different types of errors"?

Reply:
If it could do `Name '${typeName}' is not contributed by any ${serviceInterface}` or a variation of then yes, but IIRC it would end up saying `There is no ${typeName} registered for FIlterContributor` which feels rather an unhelpful message.

Reply:
I guess `Name 'MultiTenantFilter' is not contributed by any FilterContributor` is better than passing lamdas around. Will go back to a single message

Reply:
resolved with 3db3644c34d0b013fc161010e2694f2482987637